crates/ibc/src/event.rs

```diff
@@ -0,0 +1,521 @@
+    ReadFromEventAttributes as _,
+};
+use namada_events::{
+    Event, EventError, EventLevel, EventToEmit, EventTypeBuilder,
+};
+use namada_macros::BorshDeserializer;
+#[cfg(feature = "migrations")]
+use namada_migrations::*;
+use serde::{Deserialize, Serialize};
+
+pub mod types {
+    //! IBC event types.
+
+    use namada_core::ibc::core::client::types::events::UPDATE_CLIENT_EVENT;
+    use namada_events::{event_type, EventType};
+
+    use super::IbcEvent;
+
+    /// Update client.
+    pub const UPDATE_CLIENT: EventType =
+        event_type!(IbcEvent, UPDATE_CLIENT_EVENT);
+}
+
+/// IBC event kind.
+#[derive(
+    Debug,
+    Clone,
+    BorshSerialize,
+    BorshDeserialize,
+    BorshDeserializer,
+    BorshSchema,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Serialize,
+    Deserialize,
+)]
+#[repr(transparent)]
+pub struct IbcEventType(pub String);
+
+impl EventToEmit for IbcEvent {
+    const DOMAIN: &'static str = "ibc";
+}
+
+impl From<IbcEvent> for Event {
+    fn from(ibc_event: IbcEvent) -> Self {
+        let mut event = Self::new(
+            EventTypeBuilder::new_of::<IbcEvent>()
+                .with_segment(ibc_event.event_type.0)
+                .build(),
+            EventLevel::Tx,
+        );
+        #[allow(deprecated)]
+        {
+            *event.attributes_mut() =
+                ibc_event.attributes.into_iter().collect();
+        }
+        event.extend(event_domain_of::<IbcEvent>());
+        event
+    }
+}
+
+impl std::fmt::Display for IbcEventType {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl FromStr for IbcEventType {
+    type Err = std::convert::Infallible;
+
+    #[inline(always)]
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(IbcEventType(s.to_owned()))
+    }
+}
+
+impl std::cmp::PartialEq<String> for IbcEventType {
+    fn eq(&self, other: &String) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl std::cmp::PartialEq<str> for IbcEventType {
+    fn eq(&self, other: &str) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl std::cmp::PartialEq<&str> for IbcEventType {
+    fn eq(&self, other: &&str) -> bool {
+        self.0.eq(other)
+    }
+}
+
+/// Wrapped IbcEvent
+#[derive(
+    Debug,
+    Clone,
+    BorshSerialize,
+    BorshDeserialize,
+    BorshDeserializer,
+    BorshSchema,
+    PartialEq,
+    Eq,
+    Serialize,
+    Deserialize,
+)]
+pub struct IbcEvent {
+    /// The IBC event type
+    pub event_type: IbcEventType,
+    /// The attributes of the IBC event
+    pub attributes: HashMap<String, String>,
+}
+
+fn validate_ibc_event_type(
+    namada_event: &Event,
+) -> Result<IbcEventType, EventError> {
+    if namada_event.kind().domain() != IbcEvent::DOMAIN {
+        return Err(EventError::InvalidEventType);
+    }
+
+    let event_type = namada_event.kind().sub_domain();
+
+    // if !matches!(
```

Comment:
what's going on here?

Reply:
I'm not sure if you're referring to the let expr, or the commented `matches!()`. the let expr is to strip the `ibc/` prefix from the event type, which `IbcEvent` instances don't contain. the `matches!()` was to check if the sub-domain was valid, but I'm not sure which ones are currently in use in namada...